### PR TITLE
Add password reset API

### PR DESCRIPTION
**Pull request by linuxdaemon**
This has been tested on the devnet by me

### CODE REVIEW

File: API.md [hunk 1]
```diff
@@ -90,3 +90,65 @@ _Note: The `message` field is meant only to describe the error to a user, this t
...
+- `email`
+
+#### Responses
+##### Success
+
+`{"status":"ok"}`
+
+#### Errors
+
+`{"error":{"id":"no_account","message":"Unable to find matching account"},"status":"error"}`
```
Comment (A-UNDERSCORE-D) [inline API.md]: Password reset tools should not expose either that an account exists or that the email or otherwise was (in)correct. Doing so exposes an attack vector for finding valid accounts.

File: API.md [hunk 2]
```diff
@@ -90,3 +90,65 @@ _Note: The `message` field is meant only to describe the error to a user, this t
...
+##### Success
+
+`{"status":"ok","password_set":true|false,"account":"<ACCOUNT_NAME>","verified":true|false,"session":"<SESSION_ID>"}`
+
+- `password_set` - Was the password changed in this request or is a future request to `/user/set/password` required
+- `verified` - Is the account a verified account or not
+
+##### Errors
+
+`{"error":{"id":"no_account","message":"Unable to find matching account"},"status":"error"}`
```
Comment (A-UNDERSCORE-D) [inline API.md]: `no_account` and `no_error` should simply be `error` with "unable to complete request". Expired code is fine as it already verified that the user has the code

File: API.md [hunk 3]
```diff
@@ -90,3 +90,65 @@ _Note: The `message` field is meant only to describe the error to a user, this t
...
+This has two main uses:
+- Simply confirming the reset request and logging in the user, which will then mark the account as unconfirmed until the user changes their password, or
+- Confirming the reset request and also setting the password to `newpass` in one request (requires sending the `newpass` parameter
+
+This allows finer control from the API consumer side as to the separation of authentication and modification of the account
+
+#### Params
+- `account` - Account name to reset the password for
+- `code` - Confirmation code from the user
+- `newpass` _optional_ - If set, sets the new password in this request, otherwise a later request to `/user/set/password` is required.
```
Comment (A-UNDERSCORE-D) [inline API.md]: newpass should not be optional, as it provides the ability for a user to get locked out.

File: API.md [hunk 4]
```diff
@@ -90,3 +90,65 @@ _Note: The `message` field is meant only to describe the error to a user, this t
...
+
+#### Responses
+##### Success
+
+`{"status":"ok","password_set":true|false,"account":"<ACCOUNT_NAME>","verified":true|false,"session":"<SESSION_ID>"}`
+
+- `password_set` - Was the password changed in this request or is a future request to `/user/set/password` required
+- `verified` - Is the account a verified account or not
```
Comment (A-UNDERSCORE-D) [inline API.md]: why does the status of verification need to be sent in this?

File: API.md [hunk 5]
```diff
@@ -90,3 +90,65 @@ _Note: The `message` field is meant only to describe the error to a user, this t
...
+#### Responses
+##### Success
+
+`{"status":"ok"}`
+
+##### Errors
+
+`{"error":{"id":"no_login","message":"You are not logged in to an account"},"status":"error"}`
+
+`{"error":{"id":"invalid_password","message":"That password is invalid"},"status":"error"}`
```
Comment (A-UNDERSCORE-D) [inline API.md:140]: see above regarding exposing information about existing accounts. Unless this only hits users that are logged in. In which case further information about why their selected password is invalid may be prudent.

**Review by A-UNDERSCORE-D — CHANGES_REQUESTED** [review]
Following the rest of the comments, this should be verified by someone else who knows C++ better than I do.

File: API.md (hunk 3 above)
Comment (linuxdaemon) [reply API.md]: No it doesn't, the user can still log in with the old password or just do a new reset request.

File: API.md (hunk 4 above)
Comment (linuxdaemon) [reply API.md]: Because as stated earlier on, if the newpass isn't sent, the account is marked unverified pending the user setting the password with user/set/password, to mimic the anope functionality

File: API.md (hunk 5 above)
Comment (linuxdaemon) [reply API.md:140]: This is simply about the provided new password, and currently there isn't an easy way to tell why it was considered invalid, it behaves the same as the registration endpoint

File: API.md (hunk 2 above)
Comment (linuxdaemon) [reply API.md]: What is `no_error` ?

File: API.md (hunk 5 above)
Comment (prawnsalad) [reply API.md:140]: This is an internal API only, it doesn't matter if it exposes if the account is available or not here as it's only snoonet/irc.com that uses it

File: API.md (hunk 3 above)
Comment (A-UNDERSCORE-D) [reply API.md]: Then it leaves an ineffectual API endpoint.

File: API.md (hunk 4 above)
Comment (A-UNDERSCORE-D) [reply API.md]: That does not change the fact that this does not need that information passed back.

File: API.md (hunk 1 above)
Comment (linuxdaemon) [reply API.md]: Ah I'm assuming you mean I should collapse no_account and wrong_code to something like reset_failed ?

Comment (linuxdaemon) [reply API.md]: In this part of the API (the `/resetpass` endpoint), `no_account` means that something didn't match, either the email or the username, doesn't say which

File: API.md (hunk 5 above)
Comment (A-UNDERSCORE-D) [reply API.md:140]: > This is an internal API only, it doesn't matter if it exposes if the account is available or not here as it's only snoonet/irc.com that uses it

That does not change the fact that internal APIs should still not blindly trust their counterparts. That's how internal breaches allow for lateral moves and additional access to things other than the breached component. Security and layers thereof should still exist in internal environments.

>This is simply about the provided new password, and currently there isn't an easy way to tell why it was considered invalid, it behaves the same as the registration endpoint

Thats fine. My concern was regarding exposing information to a connection that was not authed for the account its attempting to mess with.

File: API.md (hunk 1 above)
Comment (A-UNDERSCORE-D) [reply API.md]: Still informs the requester that the data does not exist, which therefore allows a search of what _does_ exist

File: API.md (hunk 2 above)
Comment (A-UNDERSCORE-D) [reply API.md]: sorry, I mistyped, I wanted `wrong_code`

File: API.md (hunk 3 above)
Comment (linuxdaemon) [reply API.md]: How? If you don't set `newpass` it just logs you in to the account without setting the password, returns the session which can be used to change the password with `/user/set/password`

File: API.md (hunk 4 above)
Comment (linuxdaemon) [reply API.md]: This is meant to behave the same as the login endpoint, but with the additional information for a password reset, since this endpoint logs the user in with the temporary code anope sends out.

File: API.md (hunk 5 above)
Comment (linuxdaemon) [reply API.md:140]: > Thats fine. My concern was regarding exposing information to a connection that was not authed for the account its attempting to mess with.

Ah, yeah that's not happening here, this endpoint requires a valid and authenticated session token

File: API.md (hunk 1 above)
Comment (linuxdaemon) [reply API.md]: Should I just "allow" it through and return status:ok for invalid data? I know some sites do that for stuff

File: API.md (hunk 2 above)
Comment (linuxdaemon) [reply API.md]: ah yeah, I agree, will merge those

File: API.md (hunk 4 above)
Comment (A-UNDERSCORE-D) [reply API.md]: Why does it need to behave the same? Its not a login endpoint. Its a password reset endpoint. Exposing information not required to complete the task at hand is a security issue. Components do not need more information than what is required to make a process happen. Also, having backend things that change based on some arbitrary behavior in an 'unrelated' API endpoint is a marker of a bad API. If you want the ability to mark a user as unverified, add an endpoint for it.

File: API.md (hunk 2 above)
Comment (linuxdaemon) [reply API.md]: I could simply merge no_account to wrong_code since technically the code is invalid for a non-existent account as well

File: API.md (hunk 1 above)
Comment (A-UNDERSCORE-D) [reply API.md]: Yeah thats what I was thinking, a "if this was correct you should get an email with further instructions"

File: API.md (hunk 2 above)
Comment (A-UNDERSCORE-D) [reply API.md]: That works

File: API.md (hunk 3 above)
Comment (A-UNDERSCORE-D) [reply API.md]: Yeah but what happens if the user never actually sets it from then? you have an account without a password. Or the session is lost for $InsertReasonHere? how do you fix it then? another reset? that seems like a whole lot of effort

Comment (linuxdaemon) [reply API.md]: The passsword isn't cleared, it just stays as the old password

File: API.md (hunk 4 above)
Comment (linuxdaemon) [reply API.md]: I may just remove the ability to not set the password at that point and just keep the set/password endpoint for changing the password in a normal logged in account

File: API.md (hunk 3 above)
Comment (A-UNDERSCORE-D) [reply API.md]: So yes, ineffectual API endpoint.

File: API.md (hunk 1 above)
Comment (prawnsalad) [reply API.md]: This is an internal API, it should absolutely not mask information. It is up to how we present the data to the user to mask the information - which this message is not used for.

@A-UNDERSCORE-D I understand that you want this to be air tight in 100% of all use cases and that people on our own team should not be trusted. But this is an extreme viewpoint that only causes issues when it comes to debugging issues further down the line internally. We must be pragmatic about these things.

File: API.md (hunk 2 above)
Comment (linuxdaemon) [reply API.md]: Done in 1fd4bdfd112a337eb759f82cf4bf24e18d373acb

File: API.md (hunk 1 above)
Comment (A-UNDERSCORE-D) [reply API.md]: @prawnsalad All APIs regardless of location should not trust their clients. What happens if this is exposed to the internet at some point? then what? "oops sorry we provided userdata to the internet"? Debugging should not occur over an API. These things log locally for a reason. There's a reason, for example, HTTP servers (including only internally accessible ones) generally return `500 Internal Server Error` and not `500 Line x in file y broke because z`. In general security through obscurity (ie, "its an internal API") is a very dangerous thing to rely on. 

>I understand that you want this to be air tight in 100% of all use cases and that people on our own team should not be trusted
>We must be pragmatic about these things.

So, in other words "security is second to my ability to debug production systems in insecure ways"?

I see no reason to not trust your team. That's fine. However the things using this API are not your team, they're code. Code is not infallible, in fact, code in general almost always has bugs.

Bottom line: If an API fails to behave as expected, it should be tested in a clean environment. If your concern is debugging, perhaps it should be unit tested so bugs don't crop up, and when and if bugs do crop up, a test network is more than sufficient to provide testing. A common practice in network security is to write code et al. as if you are already compromised, and want to mitigate damage. This is also known as [defensive development](https://en.wikipedia.org/wiki/Defensive_programming). These practices should be applied to anything that requires high security (ie, works in any way shape or form with userdata), lest the GDPR come to bite us. 

This is not me trying to be difficult, this is not me being extreme. Me being extreme would be me pushing that every single server be connected over a VPN and this API (with EVERY security consideration I've already mentioned in place) only accessible via that VPN AND through firewalls on the specific machine. Security of thousands of users is not something to be taken lightly. While that would be nice, it would also be very difficult to bring existing networks into.

File: API.md (hunk 4 above)
Comment (linuxdaemon) [reply API.md]: Resolved by b5186745fa7d0415ff92250d613457e3024bbcc7

File: API.md (hunk 3 above)
Comment (linuxdaemon) [reply API.md]: Resolved by b5186745fa7d0415ff92250d613457e3024bbcc7

File: API.md (hunk 1 above)
Comment (linuxdaemon) [reply API.md]: Resolved by 186bc64e99c05166c6cee5361ba8539d992108cc
@prawnsalad For debugging, the specific error is logged internally and can be checked there.

File: API.md (hunk 4 above)
Comment (A-UNDERSCORE-D) [reply API.md]: That sounds pretty clean to me

File: API.md (hunk 1 above)
Comment (prawnsalad) [reply API.md]: re. VPNs and firewalls, this is something that I have already put onto the todo list since nobody in previous years seems to have considered that enough to put it in place.

re. debugging, this isn't about putting debugging into API. This is about putting relevant debugging information into the correct places as not to waste peoples time. If people start complaining that they cannot login, then if irc.com devs cannot see generic errors, they then have to go a level deeper and ask network level staff for more logs. You may not care about this, but we do as this is when it starts costing money, time, and can even delay debugging if everybody is not around. This is not just our use case - this is a very common thing to bare in mind for any project.

I fully understand the angle you are coming from, but you must also understand that a volunteer non-commercial project has entirely different expectations from a commercial project. You may not care so much for debugging a production system when random issues happen that may not appear during a staging test, but I and future organisations using the network do.

Comment (A-UNDERSCORE-D) [reply API.md]: >then if irc.com devs cannot see generic errors, they then have to go a level deeper and ask network level staff for more logs

Why is this an issue? especially when, if you provide errors like "this password doesn't match this account", its a very nice target for bruteforcing. If the API is exposed globally, even more so. The data is there for IRC.com devs. You want us to work together, you keep saying so, so why cant we?

Also, if me, as a volunteer for a "volunteer non-commercial project" has tighter security requirements than you, a CEO of a commercial project which will handle "many more users", then I think you have a problem.

Comment (A-UNDERSCORE-D) [reply API.md]: also:
>You may not care so much for debugging a production system when random issues happen that may not appear during a staging test, but I and future organisations using the network do.

I do care. There are provisions for debugging. They're just not exposed over the API.

Comment (A-UNDERSCORE-D) [reply API.md]: >This is not just our use case - this is a very common thing to bare in mind for any project.

Not in any project that I've seen. I'm pushing for industry standard security practices.

Comment (prawnsalad) [reply API.md]: > Why is this an issue? especially when, if you provide errors like "this password doesn't match this account", its a very nice target for bruteforcing. If the API is exposed globally, even more so.

Again, this is an internal API. We won't be brute forcing our own users. This API is never going to be made public as it's an internal API. I'm not sure how this can be made any more clear.

> The data is there for IRC.com devs. You want us to work together, you keep saying so, so why cant we?

I'm not sure what you're saying... the relevant snoonet staff have access to the same data if not more.

> Also, if me, as a volunteer for a "volunteer non-commercial project" has tighter security requirements than you, a CEO of a commercial project which will handle "many more users", then I think you have a problem.

You are not understanding the differences between the security of an internal system VS an external public accessible system. I wouldn't go around suggesting people have problems when you don't have the full knowledge of both sides of the argument. There are individual teams, web and network. Web teams don't generally have access to network related logs. However, if logs are failing then it's the web team that would first get the complaints. If the web team can see via web service logs that it's an incorrect password failure, they can react on that. if they see that it's a code failure issue, they can react on that. If it's on any other issue, they can react on that. Depending on the type of error, they may invoke the network team to look further. If the error is too vague and they depend on the network team to dig into logs but the network team is not available at that time, this could be many hours wasted if it's a simple error that could have been handled already.

This is how working with multiple internal teams work. Limiting the data - sure. Proving just enough data to help debugging between teams goes a long way.

***Again - this is all internal data.*** irc.com *and snoonet staff* already know if an account exists or not. Providing this information for debugging is not leaking anything.

Comment (prawnsalad) [reply API.md]: Also, lets be honest. This API, internal or not, is more secure and privacy focussed than what's publicly available to anybody out there.

> NickServ(NickServ@services)- Nick wedwefgrtyhj isn't registered.

Comment (A-UNDERSCORE-D) [reply API.md]: >Again, this is an internal API. We won't be brute forcing our own users. This API is never going to be made public as it's an internal API. I'm not sure how this can be made any more clear.

It exists on the internet. It can be attacked. Yes its firewalled and authenticated. That. Doesnt. Matter. The auth code could have a bug in it. The python frontend could be broken. Your backend could be broken. the API keys could be stolen. This happens. Ask any number of compromised software

In fact, the fact that it exists on the internet, and is *designed* to be used as an external API, means that it is **not** an internal API. It simply currently is only accessed by internal people and code. (which again, can be compromised)

_**Act as if you are already compromised and working to mitigate an attack**_
This doesnt just apply to external code, in fact, it can NOT just apply external code. It must apply to the entire infrastructure.

>I'm not sure what you're saying... the relevant snoonet staff have access to the same data if not more.

You said, and I quote:
>...then if irc.com devs cannot see generic errors, they then have to go a level deeper and ask network level staff for more logs.

Hence the question regarding working together.

>You are not understanding the differences between the security of an internal system VS an external public accessible system.

Internal systems should be as secure if not more secure. Also. **It is on the internet** and even IF it was on an internal network that was 100% airgapped. Malicious code and parties exist. 

>Web teams don't generally have access to network related logs. However, if logs are failing then it's the web team that would first get the complaints. If the web team can see via web service logs that it's an incorrect password failure, they can react on that. if they see that it's a code failure issue, they can react on that. If it's on any other issue, they can react on that. Depending on the type of error, they may invoke the network team to look further. If the error is too vague and they depend on the network team to dig into logs but the network team is not available at that time, this could be many hours wasted if it's a simple error that could have been handled already.

A extra time is worth not exposing what could possibly release user data.

>Also, lets be honest. This API, internal or not, is more secure and privacy focussed than what's publicly available to anybody out there.
>>NickServ(NickServ@services)- Nick wedwefgrtyhj isn't registered.

IRC already has existing logging, flood protections, and monitoring. I don't see how this is relevant. Not to mention that `/ns info`-ing a nick isn't the same as bruteforcing password reset tokens, or passwords in general

Not to mention the fact, that NickServ itself will KILL users off after attacks like this API provides the ability to do. And the network will kill users off when flooding (see aforementioned flood protection)

# bottom line

This supposedly "internal" API, as demonstrated above, is nether internal nor intended to be internal. It just happens to be currently used by internal people and code (which, again, again, can be compromised)

And, if, and I really hope it never happens, but IF this is compromised, and leads to an investigation and fine due to GDPR and other laws regarding user privacy, do we really want to think back to this and say "perhaps we should have implemented more security"?

**Review by A-UNDERSCORE-D — COMMENTED** [review]
Okay as these changes stand I'm good with them, assuming @sadieltm is okay with the code, and assuming updated code is also tested

File: API.md (hunk 1 above)
Comment (prawnsalad) [reply API.md]: I've read, understood, and noted your comments.

Of course, if any fine or investigation is made towards Snoonet or irc.com, I am the one ultimately responsible. I've mentioned this before. I fully realise that I am also taking on the responsibility of all the previous fuckups from Snoonet staff. The good or bad.

Based on previous conversations I don't expect you nor I to see eye to eye on these topics for the most part, but we should at least respect each others input and role in the network. And one of the important factors at this stage in this whole platform is the ease of debugging and development as we push forward, which is where these error messages come into play.

Comment (prawnsalad) [reply API.md]: # bottom line x2
oops, missed this attention grabbing title in my previous comment

Comment (linuxdaemon) [reply API.md]: Ok, if you have a legitimate change to request, request changes on the diff. Otherwise I'm going to lock the thread.

**Comment by linuxdaemon** [comment]
@A-UNDERSCORE-D Ok, the new changes have been fully tested

**Review by ghost — APPROVED** [review]

**Review by A-UNDERSCORE-D — APPROVED** [review]
